tests/test_freezable_type_caster.cpp

```diff
@@ -0,0 +1,418 @@
+/*
+    tests/test_freezable_type_caster.cpp
+
+    This test verifies that const-ness is propagated through the type_caster framework.
+    The returned FreezableInt type is a custom type that is frozen when returned as a
+    const T or std::reference<const T>, and not frozen otherwise.
+    This test is somewhat complicated because it introduces a custom python type to
+    manage the frozen aspect of the type rather then relying on native pybind11 which
+    does not support such a feature.
+
+    All rights reserved. Use of this source code is governed by a
+    BSD-style license that can be found in the LICENSE file.
+*/
+#include <pybind11/pybind11.h>
+
+#include <deque>
+#include <map>
+#include <type_traits>
+#include <unordered_map>
+#include <cstring>
+#include <memory>
+
+#include "pybind11_tests.h"
+
+/// C++ type
+class FreezableInt {
+ public:
+  int64_t x = 1;
+};
+
+/// Python wrapper for C++ type which supports mutable/immutable variants.
+typedef struct PyFreezableInt {
```

Comment:
Does this need to be implemented with the raw CPython API?

Reply:
I want a freezable representation of c++ object available to python. I don't see a better option to achieve what I'm trying to achieve here.

Reply:
You could make some kind of a wrapper `FreezableIntWrapper`, expose that as `class_<FreezableIntWrapper>`, and have a type-caster specialization for `FreezableInt`, which uses the (auto-generated by pybind11) `FreezableIntWrapper`.

I think it will save you a lot of overhead in code (at the cost of being maybe slightly less efficient). I currently don't have the time to dig into it, but maybe over the next days.

Reply:
> You could make some kind of a wrapper `FreezableIntWrapper`, expose that as `class_<FreezableIntWrapper>`, and have a type-caster specialization for `FreezableInt`, which uses the (auto-generated by pybind11) `FreezableIntWrapper`.
> 
> I think it will save you a lot of overhead in code (at the cost of being maybe slightly less efficient). I currently don't have the time to dig into it, but maybe over the next days.

The new test as-is has many lines of code, but it is as straightforward as it can possibly be and is 100% representative of how real-world non-pybind11-wrapped types interact with pybind11. Your suggested alternative will be more compact, but lose those two virtues. My suggestion is to respect Laramie's decision to go this route and not spend any more time working on the test. It seems very unlikely that the extra investment will ever pay off, via reduced maintenance cost. Adjustments for future changes in the Python C API are likely to be tiny. Maintaining some fabricated wrapper code could well have a higher maintenance cost, in part simply because it's less straightforward for anyone to get their head around.

Reply:
@rwgk, you said yourself we don't support this:

> Note that the loss of `const` isn't just limited to this situation. It existed in Boost.Python already, and is very difficult to avoid. (I wouldn't try.)

As I said, if there's a straightforward way to change pybind11 such that this weird scenario is supported, we can do it. But I strongly oppose adding and maintaining more raw C API than necessary, especially in a test (I believe that's what @bstaletic complained about as well?).
There ought to be a better way to test this, without the introduction of a whole new test file and 200 lines of raw C API?

Reply:
> @rwgk, you said yourself we don't support this:
> 
> > Note that the loss of `const` isn't just limited to this situation. It existed in Boost.Python already, and is very difficult to avoid. (I wouldn't try.)

What I meant: loss of `const` is a fundamental shortcoming, but also deep, therefore I wouldn't try to fix it at that fundamental level.

> As I said, if there's a straightforward way to change pybind11 such that this weird scenario is supported, we can do it.

The scenario may be unusual, but looking at this with a birds-eye view, it's weird to have e.g. `set` and `frozenset`. Laramie's test is reflective of the real-world situation he wrote this test for (related to Google's OSS protobuf library), which will implement the freezing logic at runtime. I think it's totally worth supporting that, with the tiny tweak he needs here.

Since the core change doesn't break any existing test, and it is tiny, would you accept it without additional test?

The only thing lost from the protobuf library viewpoint is a protection against regression, but that's a small loss (we'd have to upstream breakage reports, probably rare).